File: max_twin_sum_of_a_linked_list_2130.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def print_linked_list(head):
    current = head
    while current:
        print(current.val, end=" -> ")
        current = current.next
    print("None")


class Solution:
    def __init__(self):
        self.stack = []
        self.maxStack = 0

    def pairSum(self, head) -> int:
        slow = fast = head

        while fast and fast.next:
            self.stack.append(slow.val)
            slow = slow.next
            fast = fast.next.next

        while slow:
            self.maxStack = max(self.maxStack, slow.val+self.stack.pop())
            slow = slow.next

        return self.maxStack


head = ListNode(1, ListNode(2, ListNode(3, ListNode(4))))
solution = Solution()
solution.pairSum(head)

File: max_twin_sum_of_a_linked_list_2130.js
function pairSum(head) {
  let slow = head;
  let fast = head;

  const stack = [];

  while (fast && fast.next) {
    stack.push(slow.val);
    slow = slow.next;
    fast = fast.next.next;
  }

  let maxTwinSum = 0;

  while (slow) {
    maxTwinSum = Math.max(maxTwinSum, stack.pop() + slow.val);
    slow = slow.next;
  }

  return maxTwinSum;
}
